tests/test_sqlPreFlight.py

```diff
@@ -0,0 +1,240 @@
+        registry.registerDatasetType(calexpType)
+
+        # add pre-existing datasets
+        for exposure in (100, 101, 110, 111):
+            for sensor in (1, 2, 3):
+                # note that only 3 of 5 sensors have datasets
+                dataId = dict(camera='DummyCam', exposure=exposure, sensor=sensor)
+                registry.addDataset(rawType, dataId=dataId, run=run)
+
+        # with empty expression
+        headers, rows = self.preFlight.selectDataUnits(collections=["test"], expr="",
+                                                       neededDatasetTypes=[rawType],
+                                                       futureDatasetTypes=[calexpType])
+        # order of columns is not defined
+        self.assertCountEqual(headers, (('Camera', 'camera'),
+                                        ('Sensor', 'sensor'),
+                                        ('Exposure', 'exposure'),
+                                        ('Visit', 'visit')))
+        rows = list(rows)
+        self.assertEqual(len(rows), 4*3)   # 4 exposures times 3 sensors
+
+        # limit to single visit
+        headers, rows = self.preFlight.selectDataUnits(collections=["test"],
+                                                       expr="Visit.visit = 10",
+                                                       neededDatasetTypes=[rawType],
+                                                       futureDatasetTypes=[calexpType])
+        self.assertEqual(len(headers), 4)  # headers must be the same
+        rows = list(rows)
+        self.assertEqual(len(rows), 2*3)   # 2 exposures times 3 sensors
+
+        # expression excludes everyhting
+        headers, rows = self.preFlight.selectDataUnits(collections=["test"],
+                                                       expr="Visit.visit = 10 and Sensor.sensor > 1",
+                                                       neededDatasetTypes=[rawType],
+                                                       futureDatasetTypes=[calexpType])
+        self.assertEqual(len(headers), 4)  # headers must be the same
+        rows = list(rows)
+        self.assertEqual(len(rows), 2*2)   # 2 exposures times 2 sensors
+
+        # expression excludes everyhting
+        headers, rows = self.preFlight.selectDataUnits(collections=["test"],
+                                                       expr="Visit.visit > 1000",
+                                                       neededDatasetTypes=[rawType],
+                                                       futureDatasetTypes=[calexpType])
+        self.assertEqual(len(headers), 4)  # headers must be the same
+        rows = list(rows)
+        self.assertEqual(len(rows), 0)
+
+        # Selecting by PhysicalFilter, this should be done via
+        # PhysicalFilter.physical_filter but that is not supported currently,
+        # instead we can do Visit.physical_filter (or Exposure.physical_filter)
```

Comment:
I don't see where this limitation comes from in the implementation.

Reply:
`PhysicalFilter` is not in the units for the `RAW` DatasetType that I defined above, so PhysicalFilter table does not appear in generated SQL.  I will need to analyze user expression to see that PhysicalFilter DataUnit is there and then add a join for that table. For that I'll need user expression to be something better than dumb string, something I plan to do later.

Reply:
Ah, this gets at the point above about dependent DataUnits (though not required ones).  "raw" does depend on a DataUnit that has a foreign key to Physical fitness, and that's another way it could be brought in rather than via the expression.